File: java/all-your-base/src/main/java/BaseConverter.java
import java.util.ArrayList;
import java.util.List;

class BaseConverter {
    private int num;

    BaseConverter(int base, int[] digits) {
        if (base < 2) {
            throw new IllegalArgumentException("Bases must be at least 2.");
        }

        this.num = 0;
        int pow = 1;
        for (int i = digits.length - 1; i >= 0; --i) {
            if (digits[i] < 0) {
                throw new IllegalArgumentException("Digits may not be negative.");
            } else if (digits[i] >= base) {
                throw new IllegalArgumentException("All digits must be strictly less than the base.");
            }
            this.num += digits[i] * pow;
            pow *= base;
        }
    }

    int[] convertToBase(int base) {
        if (base < 2) {
            throw new IllegalArgumentException("Bases must be at least 2.");
        }

        List<Integer> result = new ArrayList<>();
        int num = this.num;
        while (num > 0) {
            result.add(0, num % base);
            num /= base;
        }
        if (result.isEmpty()) {
            result.add(0);
        }
        return result.stream().mapToInt(i -> i).toArray();
    }
}

File: kotlin/all-your-base/src/main/kotlin/BaseConverter.kt
class BaseConverter(val inBase: Int, val inDigits: IntArray) {
    init {
        require(inBase >= 2) { "Bases must be at least 2." }
        require(!inDigits.isEmpty()) { "You must supply at least one digit." }
        require(inDigits.first() != 0 || inDigits.size == 1) { "Digits may not contain leading zeros." }
        require(inDigits.all { it >= 0 }) { "Digits may not be negative." }
        require(inDigits.all { it < inBase }) { "All digits must be strictly less than the base." }
    }

    val toBase10: Int by lazy {
        powersOf(inBase).zip(inDigits.reversed().asSequence()) { a, b -> a * b }.sum()
    }

    fun convertToBase(outBase: Int): IntArray {
        require(outBase >= 2) { "Bases must be at least 2." }
        
        val outDigits = mutableListOf<Int>()
        var num = toBase10
        while (num != 0) {
            outDigits.add(num % outBase)
            num = num / outBase
        }
        if (outDigits.isEmpty()) {
            outDigits.add(0)
        }

        return outDigits.reversed().toIntArray()
    }

    private fun powersOf(num: Int): Sequence<Int> = generateSequence(1) { it * num }
}
